File: getldrfiles.py
# CurseClient
# Made by ObjectiveMoon
import re
from urllib.parse import quote
headers = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:148.0) Gecko/20100101 Firefox/148.0",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9",
    "Sec-GPC": "1",
    "Upgrade-Insecure-Requests": "1",
    "Sec-Fetch-Dest": "document",
    "Sec-Fetch-Mode": "navigate",
    "Sec-Fetch-Site": "none",
    "Sec-Fetch-User": "?1",
    "Priority": "u=0, i",
}
async def parser(data, base_url):
    files = []
    match = list(re.finditer(r'<div class=" file-row">(.*?)(?=<div class=" file-row">|</div><div class="files-results-bar files-results-bar--bottom")', data, re.DOTALL))
    if match:
        for match in match:
            row = match.group(1)
            fid = re.search(r'file-details-(\d+)', row)
            name = re.search(r'class="name" title="([^"]+)"', row)
            download = re.search(r'/download/(\d+)', row)
            version_matches = re.findall(r'class="file-row-chip(?: file-row-chip-more)?" title="([^"]+)"', row)
            loader = re.search(r'class="ellipsis">(Forge|Fabric|NeoForge|Quilt)</span>', row)
            size = re.search(r'<div class="row-detail"><div class=" tooltip-wrapper"><span>([^<]+(?:KB|MB|GB))</span>', row)
            downloads = re.search(r'class="row-detail downloads".*?<span class="ellipsis">([^<]+)</span>', row, re.DOTALL)
            uploaded = re.search(r'class="row-detail".*?<span>([A-Z][a-z]{2} \d{1,2}, \d{4})</span>', row, re.DOTALL)
            fid = (fid or download)
            if not fid or not name:
                continue
            files.append({
                "filename": "",
                "versions": version_matches,
                "loaders": [loader.group(1)] if loader else [],
                "uploaded": uploaded.group(1) if uploaded else "",
                "size": size.group(1) if size else "",
                "downloads": downloads.group(1) if downloads else "",
                "fileurl": f"https://www.curseforge.com{base_url[len('https://www.curseforge.com'):].rstrip('/')}/files/{fid.group(1)}",
                "displayname": name.group(1),
            })
        return files

async def get_ldrfiles(session, base_url, page, page_size, headers):
    url = f"{base_url}/files/all?page={page}&pageSize={page_size}&showAlphaFiles=hide"
    resp = await session.get(url, headers=headers)
    data = await resp.text()
    nums = re.findall(r'<li[^>]*>\s*<button>(\d+)</button>\s*</li>', data)
    if not nums:
        nums = re.findall(r'<ul class="page-numbers">(.*?)</ul>', data, re.DOTALL)
        if nums:
            nums = re.findall(r'<button>(\d+)</button>', nums[0])
    pages = max((int(n) for n in nums), default=page)
    fmap: dict[str, str] = {}
    for m in re.finditer(r'\\"id\\":(\d+),\\"fileName\\":\\"([^\\"]+)\\"', data):
        name = m.group(2)
        if name.lower().endswith(".jar"):
            fmap[m.group(1)] = name
    if not fmap:
        for m in re.finditer(r'"id":(\d+),"fileName":"([^"]+)"', data):
            name = m.group(2)
            if name.lower().endswith(".jar"):
                fmap[m.group(1)] = name
    files = await parser(data, base_url)
    for f in files:
        if idmatch := re.search(r'/files/(\d+)$', f["fileurl"]):
            fid = idmatch.group(1)
            if fnamer := fmap.get(fid):
                f["filename"] = fnamer
        if not f["filename"]:
            f["filename"] = f.pop("displayname", "")
        else:
            f.pop("displayname", None)
    return files, pages

File: native/src/getldrfiles.rs
use crate::ModFile;
use regex::Regex;

pub async fn parser(data: &str, base_url: &str) -> Vec<ModFile> {
    let file_id_re = Regex::new(r#"file-details-(\d+)"#).unwrap();
    let name_re = Regex::new(r#"class="name" title="([^"]+)""#).unwrap();
    let download_re = Regex::new(r#"/download/(\d+)"#).unwrap();
    let version_re =
        Regex::new(r#"class="file-row-chip(?: file-row-chip-more)?" title="([^"]+)""#).unwrap();
    let loader_re = Regex::new(r#"class="ellipsis">(Forge|Fabric|NeoForge|Quilt)</span>"#).unwrap();
    let size_re = Regex::new(
        r#"<div class="row-detail"><div class=" tooltip-wrapper"><span>([^<]+(?:KB|MB|GB))</span>"#,
    )
    .unwrap();
    let downloads_re =
        Regex::new(r#"class="row-detail downloads".*?<span class="ellipsis">([^<]+)</span>"#)
            .unwrap();
    let uploaded_re =
        Regex::new(r#"class="row-detail".*?<span>([A-Z][a-z]{2} \d{1,2}, \d{4})</span>"#).unwrap();
    let base = base_url.trim_end_matches('/');

    data.split(r#"<div class=" file-row">"#)
        .skip(1)
        .filter_map(|row| {
            let row = row
                .split(r#"</div><div class="files-results-bar files-results-bar--bottom"#)
                .next()
                .unwrap_or(row);
            let id = file_id_re
                .captures(row)
                .or_else(|| download_re.captures(row))?;
            let name = name_re.captures(row)?;
            Some(ModFile {
                filename: name[1].to_owned(),
                versions: version_re
                    .captures_iter(row)
                    .map(|m| m[1].to_owned())
                    .collect(),
                loaders: loader_re
                    .captures(row)
                    .map(|m| vec![m[1].to_owned()])
                    .unwrap_or_default(),
                uploaded: uploaded_re
                    .captures(row)
                    .map(|m| m[1].to_owned())
                    .unwrap_or_default(),
                size: size_re
                    .captures(row)
                    .map(|m| m[1].to_owned())
                    .unwrap_or_default(),
                downloads: downloads_re
                    .captures(row)
                    .map(|m| m[1].to_owned())
                    .unwrap_or_default(),
                fileurl: format!("{base}/files/{}", &id[1]),
            })
        })
        .collect()
}

pub async fn get_ldrfiles(
    http: &reqwest::Client,
    base_url: &str,
    page: u32,
    page_size: u32,
) -> Result<(Vec<ModFile>, u32), reqwest::Error> {
    let url = format!(
        "{}/files/all?page={page}&pageSize={page_size}&showAlphaFiles=hide",
        base_url.trim_end_matches('/')
    );
    let data = http.get(url).send().await?.text().await?;
    let page_re = Regex::new(r#"<li[^>]*>\s*<button>(\d+)</button>\s*</li>"#).unwrap();
    let pages = page_re
        .captures_iter(&data)
        .filter_map(|m| m[1].parse().ok())
        .max()
        .unwrap_or(page);
    let file_re = Regex::new(r#"\\?"id\\?":(\d+),\\?"fileName\\?":\\?"([^\\"]+)\\?""#).unwrap();
    let fallback_re = Regex::new(r#""id":(\d+),"fileName":"([^"]+)""#).unwrap();
    let mut names = std::collections::HashMap::new();
    for captures in file_re
        .captures_iter(&data)
        .chain(fallback_re.captures_iter(&data))
    {
        if captures[2].to_ascii_lowercase().ends_with(".jar") {
            names.insert(captures[1].to_owned(), captures[2].to_owned());
        }
    }
    let files = parser(&data, base_url)
        .await
        .into_iter()
        .map(|mut file| {
            let id = file.fileurl.rsplit('/').next().unwrap_or_default();
            if let Some(filename) = names.get(id) {
                file.filename = filename.clone();
            }
            file
        })
        .collect();
    Ok((files, pages))
}
